modules/standard/IO.chpl

```diff
@@ -2311,6 +2311,14 @@ private proc defaultSerializeType(param writing : bool) type {
   else return defaultDeserializer;
 }
 
+@chpldoc.nodoc
+proc isDefaultSerializerType(type t) param : bool {
```

Comment:
I agree that would be better. I don't think we have syntax for defining custom interfaces yet though. Is that right?

Reply:
It looks like [ChapelHashing](https://github.com/chapel-lang/chapel/blob/main/modules/internal/ChapelHashing.chpl#L28) defines an interface in module level code

Reply:
But you could very easily run into issues with the implementation, given that it isn't complete

Reply:
Okay nice, I'll give that a shot.

Reply:
Thinking about this some more, I'm not sure if it makes sense to create an interface for this particular context: `defaultSerializer` and `precisionSerializer` both have the same exact interface as the other serializers. The only thing that's special/different about them is that some of the built in types (ranges, domains, arrays) handle serialization in a special manner when the fileWriter has a default serializer — this isn't a property of the type's interface, just its identity. Furthermore, the special-casing logic is something of an implementation detail that could also be handled in a number of other ways, so I'd be wary of adding a public interface to the IO module for this purpose if we were going to (for example) refactor things to move the special formatting of those types into the serializers themselves instead of their current locations (undocumented procs in ChapelIO, etc) at some point in the future. Please lmk if you disagree.

Reply:
> this isn't a property of the type's interface, just its identity.

It's not currently a property of the type's interface, but it could become one if we felt that doing so was useful.  After all, checks like `isNumericType` relate to the identity of ints, rather than the presence of specific operators on it.

> Furthermore, the special-casing logic is something of an implementation detail that could also be handled in a number of other ways, so I'd be wary of adding a public interface to the IO module for this purpose if we were going to (for example) refactor things to move the special formatting of those types into the serializers themselves instead of their current locations (undocumented procs in ChapelIO, etc) at some point in the future.

I think it's fair to avoid making the interface public.  It's also fair to argue that means it shouldn't exist as an option at all - even if we were to add an undocumented interface, using abstraction in this way means that other serializers could be written that also used that interface.  That could lead to hijacking of our IO output handling.  Your current strategy has the benefit of limiting the impact to only supporting serializers we explicitly list ourselves in this way.  Someone would have to be using a compromised version of Chapel's IO module in order to run into problems, whereas an interface approach would open us up to compromise at the addition of any particular package.

So I think I agree that interfaces is not the way to go here.  However, it does still seem unfortunate to add an impact to compile-time and code size for default programs that don't otherwise use the new serializer as a result of this patch.  Can you think of other ways to avoid that?

Reply:
> However, it does still seem unfortunate to add an impact to compile-time and code size for default programs that don't otherwise use the new serializer as a result of this patch. Can you think of other ways to avoid that?

I wrestled with this for a bit and didn't come up with anything. On the bright side, I had to bump the count in `test/optimizations/deadCodeElimination/elliot/countDeadModules.good`, which I think means that the module does get pruned before codegen for programs that don't use it (and therefore code sizes shouldn't increase).

Reply:
Sounds good!  If we think of another strategy, we can always do something about it later